File: apps/minecraft/velocity/src/main/java/crabcraft/net/crabUtilities/velocity/LoginStreakPublisher.java
package crabcraft.net.crabUtilities.velocity;

import com.google.gson.JsonObject;
import crabcraft.net.crabUtilities.velocity.db.LoginStreakService;
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPoolConfig;

/**
 * Mirrors login streak snapshots into Redis so backend Spigot servers
 * can render them through PlaceholderAPI without each one talking to
 * Postgres.
 *
 * <p>Two writes per recorded login:
 * <ul>
 *   <li>{@code HSET crabutilities:streaks &lt;uuid&gt; &lt;json&gt;} —
 *       authoritative cache the backends seed from on player join.</li>
 *   <li>{@code PUBLISH crabutilities:streaks-updates &lt;json&gt;} —
 *       live notification for backends to refresh their in-memory
 *       cache on the spot.</li>
 * </ul>
 *
 * <p>The published payload always includes the player UUID so a single
 * subscription handles all players.
 */
public class LoginStreakPublisher {

    public static final String HASH_KEY = "crabutilities:streaks";
    public static final String UPDATE_CHANNEL = "crabutilities:streaks-updates";

    private final CrabUtilitiesVelocity plugin;
    private final VelocityConfig config;
    private JedisPool jedisPool;
    private volatile boolean redisFailureLogged;

    public LoginStreakPublisher(CrabUtilitiesVelocity plugin, VelocityConfig config) {
        this.plugin = plugin;
        this.config = config;
        connect();
    }

    private void connect() {
        JedisPoolConfig poolConfig = new JedisPoolConfig();
        poolConfig.setMaxTotal(2);
        if (config.getRedisPassword() != null && !config.getRedisPassword().isEmpty()) {
            this.jedisPool = new JedisPool(poolConfig, config.getRedisHost(),
                    config.getRedisPort(), 2000, config.getRedisPassword());
        } else {
            this.jedisPool = new JedisPool(poolConfig, config.getRedisHost(),
                    config.getRedisPort(), 2000);
        }
        plugin.getLogger().info("Login streak publisher ready; Redis will be retried on publish if unavailable.");
    }

    public void publish(String uuid, LoginStreakService.StreakSnapshot snapshot, int resetHourUtc) {
        if (jedisPool == null || jedisPool.isClosed() || snapshot == null) return;

        long now = System.currentTimeMillis() / 1000L;
        long expiresAt = LoginStreakService.expiryOf(snapshot.lastLoginAt, resetHourUtc);
        boolean active = now < expiresAt;

        JsonObject payload = new JsonObject();
        payload.addProperty("uuid", uuid);
        payload.addProperty("current_streak", active ? snapshot.currentStreak : 0);
        payload.addProperty("pending_streak", snapshot.currentStreak);
        payload.addProperty("longest_streak", snapshot.longestStreak);
        payload.addProperty("last_login_at", snapshot.lastLoginAt);
        payload.addProperty("streak_started_at", snapshot.streakStartedAt);
        payload.addProperty("expires_at", expiresAt);
        payload.addProperty("active", active);

        String body = payload.toString();
        try (Jedis jedis = jedisPool.getResource()) {
            jedis.hset(HASH_KEY, uuid, body);
            jedis.publish(UPDATE_CHANNEL, body);
            if (redisFailureLogged) {
                plugin.getLogger().info("Login streak Redis publisher recovered.");
                redisFailureLogged = false;
            }
        } catch (Exception e) {
            if (!redisFailureLogged) {
                plugin.getLogger().warn("Failed to publish login streak for {}; will retry on later logins", uuid, e);
                redisFailureLogged = true;
            } else {
                plugin.getLogger().debug("Failed to publish login streak for {}: {}", uuid, e.getMessage());
            }
        }
    }

    public void shutdown() {
        if (jedisPool != null && !jedisPool.isClosed()) {
            try { jedisPool.close(); } catch (NoClassDefFoundError ignored) {}
            jedisPool = null;
        }
    }
}

File: apps/minecraft/velocity/src/main/kotlin/crabcraft/net/crabUtilities/velocity/LoginStreakPublisher.kt
package crabcraft.net.crabUtilities.velocity

import com.google.gson.JsonObject
import crabcraft.net.crabUtilities.velocity.db.LoginStreakService
import redis.clients.jedis.JedisPool

/**
 * Mirrors login streak snapshots into Redis so backend Spigot servers can render them through PlaceholderAPI without
 * each one talking to Postgres. Each update writes the authoritative hash and publishes its UUID-bearing payload for
 * live cache refreshes.
 */
open class LoginStreakPublisher(private val plugin: CrabUtilitiesVelocity, private val config: VelocityConfig) {
    private var jedisPool: JedisPool? = RedisPools.create(config, 2)
    @Volatile private var redisFailureLogged = false

    init {
        plugin.getLogger().info("Login streak publisher ready; Redis will be retried on publish if unavailable.")
    }

    open fun publish(uuid: String, snapshot: LoginStreakService.StreakSnapshot?, resetHourUtc: Int) {
        val pool = jedisPool
        if (pool == null || pool.isClosed || snapshot == null) return
        val now = System.currentTimeMillis() / 1000L
        val expiresAt = LoginStreakService.expiryOf(snapshot.lastLoginAt, resetHourUtc)
        val active = now < expiresAt
        val payload =
            JsonObject().apply {
                addProperty("uuid", uuid)
                addProperty("current_streak", if (active) snapshot.currentStreak else 0)
                addProperty("pending_streak", snapshot.currentStreak)
                addProperty("longest_streak", snapshot.longestStreak)
                addProperty("last_login_at", snapshot.lastLoginAt)
                addProperty("streak_started_at", snapshot.streakStartedAt)
                addProperty("expires_at", expiresAt)
                addProperty("active", active)
            }
        try {
            pool.resource.use { jedis ->
                jedis.eval(
                    PUBLISH_SCRIPT,
                    listOf(HASH_KEY),
                    listOf(uuid, snapshot.lastLoginAt.toString(), payload.toString(), UPDATE_CHANNEL),
                )
                if (redisFailureLogged) {
                    plugin.getLogger().info("Login streak Redis publisher recovered.")
                    redisFailureLogged = false
                }
            }
        } catch (error: Exception) {
            if (!redisFailureLogged) {
                plugin
                    .getLogger()
                    .warn("Failed to publish login streak for {}; will retry on later logins", uuid, error)
                redisFailureLogged = true
            } else plugin.getLogger().debug("Failed to publish login streak for {}: {}", uuid, error.message)
        }
    }

    open fun shutdown() {
        val pool = jedisPool
        if (pool != null && !pool.isClosed) {
            try {
                pool.close()
            } catch (ignored: NoClassDefFoundError) {}
            jedisPool = null
        }
    }

    companion object {
        const val HASH_KEY = "crabutilities:streaks"
        const val UPDATE_CHANNEL = "crabutilities:streaks-updates"
        private val PUBLISH_SCRIPT =
            """
            local current = redis.call('HGET', KEYS[1], ARGV[1])
            if current then
                local ok, decoded = pcall(cjson.decode, current)
                if ok and type(decoded) == 'table'
                        and (tonumber(decoded['last_login_at']) or 0) > tonumber(ARGV[2]) then
                    return 0
                end
            end
            redis.call('HSET', KEYS[1], ARGV[1], ARGV[3])
            redis.call('PUBLISH', ARGV[4], ARGV[3])
            return 1
            """
                .trimIndent() + "\n"
    }
}
